test_util/fault_injection_test_fs.h

```diff
@@ -238,14 +278,92 @@ class FaultInjectionTestFS : public FileSystemWrapper {
     error_ = io_error;
   }
 
+  // Specify what the operation, so we can inject the right type of error
+  enum ErrorOperation : char {
+    kRead = 0,
+    kOpen,
+  };
+
+  // Set thread-local parameters for error injection. The first argument,
+  // seed is the seed for the random number generator, and one_in determines
+  // the probability of injecting error (i.e an error is injected with
+  // 1/one_in probability)
+  void SetThreadLocalReadErrorContext(uint32_t seed, int one_in) {
+    struct ErrorContext* ctx =
+          static_cast<struct ErrorContext*>(thread_local_error_->Get());
+    if (ctx == nullptr) {
+      ctx = new ErrorContext(seed);
+      thread_local_error_->Reset(ctx);
+    }
+    ctx->one_in = one_in;
+    ctx->count = 0;
+  }
+
+  // Inject an error. For a READ operation, a status of IOError(), a
+  // corruption in the contents of scratch, or truncation of slice
+  // are the types of error with equal probability. For OPEN,
+  // its always an IOError.
+  IOStatus InjectError(ErrorOperation op, Slice* slice, char* scratch);
+
+  // Get the count of how many times we injected since the previous call
+  int GetAndResetErrorCount() {
+    ErrorContext* ctx =
+          static_cast<ErrorContext*>(thread_local_error_->Get());
+    int count = 0;
+    if (ctx != nullptr) {
+      count = ctx->count;
+      ctx->count = 0;
+    }
+    return count;
+  }
+
+  void EnableErrorInjection() {
```

Comment:
EnableErrorInjection is only used for read? The error is returned from InjectionError instead of GetError. In this case, we do not need to call SetFilesystemActive to set the IO Error. Can we unify them?

Reply:
They serve different purposes though. SetFileSystemActive() will set an error globally, whereas EnableErrorInjection() only does it for the calling thread. The former is difficult to use in db_stress, whereas the latter is difficult to use when you want to inject an error in, say, a background thread. So I think both are needed.

Reply:
> They serve different purposes though. SetFileSystemActive() will set an error globally, whereas EnableErrorInjection() only does it for the calling thread. The former is difficult to use in db_stress, whereas the latter is difficult to use when you want to inject an error in, say, a background thread. So I think both are needed.

I see. Thanks for the explanation.